core/pkg/artifacts/saver.go

```diff
@@ -144,99 +146,157 @@ func (as *ArtifactSaver) uploadFiles(artifactID string, manifest *Manifest, mani
+		// We shouldn't ordinarily need to retry at all: our internal client handles
+		// retryable errors, and the only failure this retry loop is for is when signed
+		// urls expire before an upload was started (exceedingly rare).
+		// Still, as long as more than half of them succeed in each iteration this will
+		// eventually terminate, so we're generous with our retry policy.
+		if len(fileSpecs) > numNeedUploading/2 {
+			return fmt.Errorf(
+				"most remaining uploads (%d/%d) have failed, giving up",
+				len(fileSpecs), numNeedUploading,
 			)
-			if err != nil {
-				return err
-			}
-			if len(fileSpecsBatch) != len(response.CreateArtifactFiles.Files.Edges) {
-				return fmt.Errorf(
-					"expected %v upload URLs, got %v",
-					len(fileSpecsBatch),
-					len(response.CreateArtifactFiles.Files.Edges),
-				)
-			}
-			// Save birth artifact ids, schedule uploads.
-			for i, edge := range response.CreateArtifactFiles.Files.Edges {
-				name := fileSpecsBatch[i].Name
-				entry := manifest.Contents[name]
-				entry.BirthArtifactID = &edge.Node.Artifact.Id
-				manifest.Contents[name] = entry
-				if edge.Node.UploadUrl == nil {
-					numDone++
-					continue
-				}
-				numInProgress++
-				task := &filetransfer.Task{
-					FileKind: filetransfer.RunFileKindArtifact,
-					Type:     filetransfer.UploadTask,
-					Path:     *entry.LocalPath,
-					Url:      *edge.Node.UploadUrl,
-					Headers:  edge.Node.UploadHeaders,
-				}
-				task.SetCompletionCallback(
-					func(t *filetransfer.Task) {
-						taskResultsChan <- TaskResult{t, name}
-					},
-				)
-				as.FileTransferManager.AddTask(task)
-			}
 		}
-		// Wait for filetransfer to catch up. If there's nothing more to schedule, wait for all in progress tasks.
-		for numInProgress > maxBacklog || (len(fileSpecsBatch) == 0 && numInProgress > 0) {
-			numInProgress--
-			result := <-taskResultsChan
-			if result.Task.Err != nil {
-				// We want to retry when the signed URL expires. However, distinguishing that error from others is not
-				// trivial. As a heuristic, we retry if the request failed more than an hour after we fetched the URL.
-				if time.Since(nameToScheduledTime[result.Name]) < 1*time.Hour {
-					return result.Task.Err
-				}
-				delete(nameToScheduledTime, result.Name) // retry
+		if len(fileSpecs) > 0 {
+			slog.Warn("some files failed to upload, retrying", "count", len(fileSpecs))
+		}
+	}
+	return nil
+}
+
+func (as *ArtifactSaver) processFiles(
+	manifest *Manifest, fileSpecs map[string]gql.CreateArtifactFileSpecInput,
+) (map[string]gql.CreateArtifactFileSpecInput, error) {
+
+	// Channels to get responses from the batch url retrievers.
+	readyChan := make(chan serverFileResponse, as.batchSize)
+	errorChan := make(chan error)
+
+	doneChan := make(chan *filetransfer.Task, as.batchSize)
+	mustRetry := map[string]gql.CreateArtifactFileSpecInput{}
+
+	numTotal := len(fileSpecs)
+	numActive := 0
+	numDone := 0
+
+	fileSpecBatches := batchify(fileSpecs, as.batchSize)
```

Comment:
since you are creating batches of `as.batchSize` and `batchFileDataRetriever` does 1 request per batch, you dont ever get any interesting parallelism, you always hover between 1 - 2 GQL requests right? E.g.:
```
t=0 - numActive = 0, you add 1 batch of 1k to the GQL pool
t=1 - numActive = 1000, nothing happens
t=2 - gql request returns, the goroutine serially processes each response and puts it on `resultChan`
t=3 - this loop processes n requests, next loop, we haven 1000-n active requests, so we add another batch.
t=4 - now we have 2000-n request active requests, a second GQL request is sent out
t=4...for a while - nothing happens until we process at least 1000 more responses. so the effective parallelism is only 2 GQL requests
```

now that I write it out, a few questions:
1. why do we want to control this parallelism? I feel like you might want independent variables for `batchSize` and number of batches in flight
2. can we just fire off all the batches? do we find gorilla actually respond slower if we overload it?

Reply:
In practice the current code spends 90% of its time with exactly 2000 outstanding requests. (it launches the second batch at 1000 because the comparison is `<=` rather than `<`). Independent control of the # of batches in flight makes sense to me, and I had thought that 5-10 batches would be better than 2, but it didn't seem to perform any faster. It didn't perform more slowly either though, so that would be fine.

In answer to both 1 & 2, we don't want to fire off all the batches and we want to control the parallelism because we don't want to end up in the situation where we've requested upload URLs far in advance of using them and letting them expire before use. Overloading Gorilla isn't really a concern.